core/src/main/scala/org/bykn/bosatsu/ToyIO.scala

```diff
@@ -0,0 +1,95 @@
+package org.bykn.bosatsu
+
+sealed abstract class ToyIO[+E, +A]
+
+object ToyIO {
+  case class Pure[A](get: A) extends ToyIO[Nothing, A]
+  case class Err[E](get: E) extends ToyIO[E, Nothing]
+  case class FlatMap[E, A, B](init: ToyIO[E, A], fn: A => ToyIO[E, B]) extends ToyIO[E, B]
+  case class RecoverWith[E, E1, A](init: ToyIO[E, A], fn: E => ToyIO[E1, A]) extends ToyIO[E1, A]
+  /**
+   * fix(f) = f(fix(f))
+   */
+  case class ApplyFix[E, A, B](arg: A, fixed: (A => ToyIO[E, B]) => (A => ToyIO[E, B])) extends ToyIO[E, B] {
```

Comment:
this constructor can't be written in Bosatsu because `ToyIO[E, B]` is in an invariant position (contravariant and covariant in the fixed function, and we can only have co-variant recursion in bosatsu. But we could define this constructor outside of bosatsu and return it via an `external def`.